bluemira/equilibria/harmonics.py

```diff
@@ -0,0 +1,682 @@
+# bluemira is an integrated inter-disciplinary design tool for future fusion
+# reactors. It incorporates several modules, some of which rely on other
+# codes, to carry out a range of typical conceptual fusion reactor design
+# activities.
+#
+# Copyright (C) 2021-2023 M. Coleman, J. Cook, F. Franza, I.A. Maione, S. McIntosh,
+#                         J. Morris, D. Short
+#
+# bluemira is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# bluemira is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+
+"""
+Spherical harmonics classes and calculations.
+"""
+from copy import deepcopy
+from enum import Enum, auto
+from typing import Dict, Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy.interpolate import RectBivariateSpline
+from scipy.special import lpmv
+
+from bluemira.base.constants import MU_0
+from bluemira.base.error import BluemiraError
+from bluemira.base.look_and_feel import bluemira_print
+from bluemira.equilibria.coils import CoilSet
+from bluemira.equilibria.equilibrium import Equilibrium
+from bluemira.equilibria.grid import Grid
+from bluemira.equilibria.plotting import PLOT_DEFAULTS
+from bluemira.geometry.coordinates import (
+    Coordinates,
+    get_area_2d,
+    get_intersect,
+    polygon_in_polygon,
+)
+from bluemira.geometry.face import BluemiraFace
+from bluemira.geometry.tools import boolean_cut, make_polygon
+
+
+def coil_harmonic_amplitude_matrix(
+    input_coils: CoilSet, max_degree: int, r_t: float
+) -> np.ndarray:
+    """
+    Construct matrix from harmonic amplitudes at given coil locations.
+
+    To get an array of spherical harmonic amplitudes/coeffcients (A_l)
+    which can be used in a spherical harmonic approximation of the
+    vacuum/coil contribution to the polodial flux (psi) do:
+        A_l = matrix harmonic amplitudes @ vector of coil currents
+
+    A_l can be used as contraints in optimisation, see spherical_harmonics_constraint.
+
+    N.B. for a single filament (coil):
+
+        A_l =  1/2 * mu_0 * I_f * sin(theta_f) * (r_t/r_f)**l *
+                    ( P_l * cos(theta_f) / sqrt(l*(l+1)) )
```

Comment:
To render maths with latex can you use https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-math for multiline or https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-math for inline/ single line

something like
```suggestion
    .. math::
        A_l =  1/2  mu_0  I_f  sin(theta_f)  (r_t/r_f)^l 
                    ( P_l  cos(theta_f) / sqrt{l(l+1)} )
```
probably needs to be tested though and there are a few other warnings in the documentation that we need to fix. You can build the docs with:

```bash
sphinx-build documentation/source  documentation/build
# to view the result as a proper web page
python -m http.server
```